packages/openapi-v3/src/__tests__/integration/controller-spec.integration.ts

```diff
@@ -203,6 +207,131 @@ describe('controller spec', () => {
     });
   });
 
+  it('allows operations to provide definition of referenced models through #/components/schema', () => {
+    class MyController {
+      @get('/todos', {
+        responses: {
+          '200': {
+            description: 'Array of Category model instances',
+            content: {
+              'application/json': {
+                schema: {
+                  $ref: '#/components/schemas/Todo',
+                  definitions: {
+                    Todo: {
+                      title: 'Todo',
+                      properties: {
+                        title: {type: 'string'},
+                      },
+                    },
+                  },
+                },
+              },
+            },
+          },
+        },
+      })
+      async find(): Promise<object[]> {
+        return []; // dummy implementation, it's never called
+      }
+    }
+
+    const spec = getControllerSpec(MyController);
+    const opSpec: OperationObject = spec.paths['/todos'].get;
+    const responseSpec = opSpec.responses['200'].content['application/json'];
+    expect(responseSpec.schema).to.deepEqual({
+      $ref: '#/components/schemas/Todo',
+    });
+
+    const globalSchemas = (spec.components || {}).schemas;
+    expect(globalSchemas).to.deepEqual({
+      Todo: {
+        title: 'Todo',
+        properties: {
+          title: {
+            type: 'string',
+          },
+        },
+      },
+    });
+  });
+
+  it('allows operations to provide definition of referenced models through #/definitions', () => {
+    class MyController {
+      @get('/todos', {
+        responses: {
+          '200': {
+            description: 'Array of Category model instances',
+            content: {
+              'application/json': {
+                schema: {
+                  $ref: '#/definitions/Todo',
+                  definitions: {
+                    Todo: {
+                      title: 'Todo',
+                      properties: {
+                        title: {type: 'string'},
+                      },
+                    },
+                  },
+                },
+              },
+            },
+          },
+        },
+      })
+      async find(): Promise<object[]> {
+        return []; // dummy implementation, it's never called
+      }
+    }
+
+    const spec = getControllerSpec(MyController);
+    const opSpec: OperationObject = spec.paths['/todos'].get;
+    const responseSpec = opSpec.responses['200'].content['application/json'];
+    expect(responseSpec.schema).to.deepEqual({
+      $ref: '#/definitions/Todo',
+    });
+
+    const globalSchemas = (spec.components || {}).schemas;
+    expect(globalSchemas).to.deepEqual({
+      Todo: {
+        title: 'Todo',
+        properties: {
+          title: {
+            type: 'string',
+          },
+        },
+      },
+    });
+  });
+
```

Comment:
Can you add a test case to allow `@api` at controller class level to supply `definitions` so that they can be shared by multiple methods in the controller?

Reply:
was talking with @nabdelgadir , I think the `@api()` decorator takes in a whole OpenAPI spec described by interface [`ControllerSpec`](https://github.com/strongloop/loopback-next/blob/master/packages/openapi-v3/src/controller-spec.ts#L29), so if a user wants to provide the model schema, the spec should be sth like:
```
{
  components: {
    schemas: {
      Todo: {
        // The Todo schema
      }
    }
  },
  paths: {
    // some path spec refers to the Todo schema provided in the root level property 
    // `components`, instead of providing the definition in itself.
  }
}
```

Reply:
I added a test in [this commit](https://github.com/strongloop/loopback-next/pull/2898/commits/e738d63d1a10a97d33d5d47b9c0fe4288fbd3b91), although I'm not sure it's what you meant. I thought that OpenAPI spec had shared definitions in `#/components/schemas`? Can you please provide an example of what you meant?

Reply:
@jannyHou Good point. We already have the ability to define `schema` at `@api` level. With that in mind, what's the additional value to allow `definitions`? Is it just for full support of json schema?

Reply:
@raymondfeng I think `definitions` is the way to add schema reference in the method level metadata, the reason is explained in Miroslav's comment: https://github.com/strongloop/loopback-next/issues/2629#issuecomment-484533465. 
`@api()` doesn't have that limitation :)

Reply:
Schema references can be used at method level regardless where the target schema is provided. With this PR, we allow the target schema to be defined in two ways:

1. At `@api` level - inside `components.schemas.<target>`
2. Use `definitions` with `$ref`:
```ts
schema: {
  $ref: #/definitions/Todo,
  definitions: {/* ... */} // [raymond] I assume the schema definitions are going to be visible to other methods?
}
```

Reply:
Since the `definition` will become a part of `#/components/schemas`, they should be visible to other methods.